---
#**CAMADA DE TRANSPORTE: PROTOCOLOS TCP, UDP E MULTIPLEXAÇÃO**

####**Relatório Técnico Acadêmico | Redes 2026 | Colaboratory**
**Data: 23/04/2026**
<div>TCP vs UDP | Handshake | Portas Lógicas | Confiabilidade</div>

*Baseado no artigo: **Arquitetura TCP/IP: Mecanismos de Controle e Portas** (Oliveira, Nakamura. Orientado por Miranda)*

##**Introdução Teórica**

A Camada de Transporte é responsável pela comunicação lógica entre processos de aplicação em diferentes hosts. Segundo Nunes (2017), enquanto a Camada de Rede entrega pacotes entre computadores, a de Transporte garante que os dados cheguem ao processo correto. Os dois protagonistas desta camada são o **TCP**, orientado à conexão e focado na integridade, e o **UDP**, que prioriza a velocidade sem garantias de entrega. A distinção entre esses serviços ocorre através da **Multiplexação**, utilizando números de portas lógicas para segmentar o tráfego de dados de forma eficiente.

---
##**Objetivos**

 - **Demonstrar** o estabelecimento de uma conexão orientada (TCP) via Handshake.
 - **Mapear** o processo de multiplexação através da identificação de portas lógicas ativas.
 - **Comparar** a estrutura de confiabilidade do TCP com a agilidade do protocolo UDP.
 - **Validar** a comunicação fim-a-fim entre a aplicação e o servidor de transporte.

---
###**1. TCP: Garantia de Conexão (Socket Stream)**

O TCP exige um "aperto de mão" (Three-way Handshake) antes de enviar dados. Vamos simular a abertura de um socket TCP para validar a prontidão da camada de transporte.

In [3]:
import socket

def realizar_handshake(host, porta):
    try:
        # AF_INET = IPv4 | SOCK_STREAM = TCP (Orientado à conexão)
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.settimeout(3)
        s.connect((host, porta))
        print(f"[CAMADA 4] Handshake realizado com sucesso em {host}:{porta}")
        s.close()
    except Exception as e:
        print(f"[ERRO] Falha na conexão: {e}")

realizar_handshake("google.com", 80)

[CAMADA 4] Handshake realizado com sucesso em google.com:80


####**O que observar:**
<a>O sucesso da conexão indica que o TCP completou o ciclo de sincronização (SYN/ACK). Como discutido no artigo de Guimarães, este protocolo é indispensável para serviços onde a perda de dados é crítica, como na navegação Web (HTTP) e transferência de arquivos.</a>

---
###**2. Identificação de Portas e Processos**

A multiplexação permite que um único endereço IP suporte múltiplos serviços. Vamos listar as portas lógicas ativas que o sistema utiliza para direcionar os pacotes.

In [1]:
# Verificando as conexões de transporte e suas portas (Local Address)
!netstat -tuln | head -n 10

Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 172.28.0.12:6000        0.0.0.0:*               LISTEN     
tcp        0      0 127.0.0.11:45229        0.0.0.0:*               LISTEN     
tcp        0      0 172.28.0.12:9000        0.0.0.0:*               LISTEN     
tcp        0      0 127.0.0.1:44325         0.0.0.0:*               LISTEN     
tcp        0      0 127.0.0.1:3453          0.0.0.0:*               LISTEN     
tcp6       0      0 :::8080                 :::*                    LISTEN     
udp        0      0 127.0.0.11:47239        0.0.0.0:*                          


####**O que observar:**
<a>As portas (ex: 80, 443, 8080) funcionam como "apartamentos" em um prédio (o IP). Como ensina Nunes, a camada de transporte usa esses números para saber exatamente qual aplicativo deve receber a informação que chegou da rede.</a>

---
###**3. Análise de Eficiência: TCP vs UDP**

Com base no artigo técnico, consolidamos a escolha do protocolo conforme o requisito da aplicação (Confiabilidade vs Velocidade).

In [2]:
import pandas as pd

dados = {
    "Característica": ["Conexão", "Confiabilidade", "Velocidade", "Aplicações"],
    "TCP (Stream)": ["Orientado (Handshake)", "Alta (Recuperação)", "Moderada", "Web, E-mail"],
    "UDP (Datagram)": ["Não orientado", "Melhor esforço", "Altíssima", "VoIP, Games"]
}

df = pd.DataFrame(dados)
print(df.to_string(index=False))

Característica          TCP (Stream) UDP (Datagram)
       Conexão Orientado (Handshake)  Não orientado
Confiabilidade    Alta (Recuperação) Melhor esforço
    Velocidade              Moderada      Altíssima
    Aplicações           Web, E-mail    VoIP, Games


####**O que observar:**
<a>A tabela resume o "trade-off" da Camada 4. O UDP é escolhido quando o atraso (delay) é mais prejudicial que a perda de um pacote individual, enquanto o TCP é obrigatório quando a ordem e a integridade são prioridades.</a>

---
##**Conclusão**

A exploração da Unidade 3.2 de Nunes, integrada aos conceitos do nosso artigo técnico, demonstrou que a eficiência de uma rede não depende apenas do meio físico, mas da escolha inteligente do protocolo de transporte. O TCP garante a robustez necessária para a integridade dos dados, enquanto o UDP oferece a fluidez essencial para aplicações de tempo real. Compreender a multiplexação por portas e os mecanismos de controle de fluxo é fundamental para o desenvolvedor de ADS, permitindo o design de sistemas distribuídos equilibrados e de alta performance.

---
##**Referências Bibliográficas**

NUNES, Sergio Eduardo. **Redes de computadores**. Londrina: Editora e Distribuidora Educacional S.A., 2017. 200 p.

GUIMARÃES, Murilo G.; OLIVEIRA, Ruy F. **Arquitetura TCP/IP**: Mecanismos de Controle e Portas. Valinhos: Faculdade Anhanguera, 2026.

TANENBAUM, Andrew S. **Redes de Computadores**. 5. ed. São Paulo: Pearson, 2011.

---
<p align="center"><b>© 2026 Murilo G. O. e Gabriel T. O. - Acadêmicos de ADS.</b></p>

---